### PR TITLE
feat: implement model evaluation pipeline with VIF centering, SES median fallback, and kernel hyperparameter selection

### LINKED ISSUE #53: Write `part2_notebook.ipynb`
# Description

Write the complete `part2/part2_notebook.ipynb` — a notebook presenting the full real-world application pipeline on the **OASIS Longitudinal Brain MRI (Alzheimer's Disease)** dataset.

This notebook serves as an "interactive report": combining theoretical explanations, illustrative code, and result analysis in a single cohesive flow, similar to `part1/part1_notebook.ipynb` in Part 1.

The file currently contains only a single empty cell with no content.

---

## Sub-issues

- [x] #48 — OLS with variable selection
- [x] #50 — Model comparison table + residual analysis
- [x] #51 — Feature importance plot

---

# Definition of Done

- [x] #48
- [x] #50
- [x] #51
- [x] Notebook runs end-to-end without errors via `uv run jupyter nbconvert --to notebook --execute part2/part2_notebook.ipynb`
- [x] Every pipeline step has a **markdown cell explaining** the reasoning behind each decision (why Group Shuffle Split, why impute instead of drop, etc.)
- [x] At least **3 core models** are present (OLS baseline, Ridge, Lasso) and at least 1 advanced method (Kernel or Bayesian)
- [x] A **Conclusions** section written in prose, interpreting results within the medical context of the dataset

---

# Proposed Implementation

## Suggested Notebook Structure

The notebook should be divided into the following sections (using `##` markdown headings):

### 0. Setup & Imports
```python
# Standard imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import GroupShuffleSplit

# Internal modules
from part2.data_pipeline import DataPipeline
from part2.model_comparison import ModelComparison
from part2.advanced_methods import AdvancedMethods
```

### 1. Dataset Overview
- Introduce the OASIS dataset (Kaggle source, 373 rows, goal: predict `MMSE`)
- Load `part2/data/oasis_longitudinal.csv` and display `.head()`, `.info()`, `.describe()`
- Visualize the distribution of the target variable `MMSE` and a missing values heatmap

### 2. Data Preprocessing with `DataPipeline`

Explain **Group Shuffle Split** over a standard random split to prevent data leakage
(the same patient appearing in both train and test sets):

```python
gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
train_idx, test_idx = next(gss.split(X, y, groups=df["Subject ID"]))
```

Call `DataPipeline.fit_transform()` on train data and `transform()` on test data.
Explain each step: drop duplicates → impute missing values → encode categoricals → z-score scaling.

### 3. OLS Baseline
- Fit OLS via `ModelComparison.train_linear_regression()`
- Print coefficients and metrics (MAE, RMSE, R²)
- Briefly comment on the baseline result

### 4. Model Comparison: Ridge with Cross-Validation
- Use k-fold (k=5 or k=10) to select the optimal `λ` for Ridge
- Plot the **CV error curve** as a function of `λ` (log scale on x-axis)
- Fit the final model with `λ_best` and evaluate metrics on the test set

### 5. Advanced Methods (Bonus)
- Fit `KernelRidge` via `AdvancedMethods`
- Compare metrics against the models from Section 4

### 6. Comparative Results Table
```python
# Expected output
| Model           | MAE   | RMSE  | R²    |
|-----------------|-------|-------|-------|
| OLS             | ...   | ...   | ...   |
| Ridge (λ=...)   | ...   | ...   | ...   |
| Kernel Ridge    | ...   | ...   | ...   |
```

### 7. Residual Analysis (Best Model)
Call `residual_plots()` from `part1/residual_analysis.py` (or equivalent)
to generate the 4 standard plots: Residuals vs Fitted, Normal Q-Q, Scale-Location, and Residuals vs Leverage.

### 8. Feature Importance — Normalized Coefficient Plot
Plot a barplot of standardized coefficients to answer:
> *Which factors have the greatest influence on the MMSE score?*

### 9. Conclusions
Write a prose discussion covering:
- Which model performed best and why
- The significance of the most influential features in the context of Alzheimer's research
- Limitations of the analysis and potential directions for improvement

## Files Affected

| File | Role |
|------|------|
| `part2/part2_notebook.ipynb` | ✏️ **File to be written** |
| `part2/data_pipeline.py` | Already implemented — `DataPipeline` class |
| `part2/model_comparison.py` | Already implemented — OLS + Ridge with CV |
| `part2/advanced_methods.py` | Already implemented — Kernel Ridge |
| `part2/data/oasis_longitudinal.csv` | Raw dataset (373 rows) |
| `part1/residual_analysis.py` | Can be reused for `residual_plots()` |

## Additional Context

- Reference `part1/part1_notebook.ipynb` for the expected presentation style (interleaved markdown and code cells)
- Every plot must include a **title, axis labels, and a legend** per the project's coding standards
- All `random_state` values must be set to `42` for reproducibility
- Format the notebook with `Black` before committing (`uv run black part2/part2_notebook.ipynb`)

### PR DESCRIPTION
**Pull request by winterSolstice25**
## 🚀 Description
Evaluate performance of models and run model on test data
Closes #53

### CODE REVIEW

**Comment by TanKhoiTV** [comment]
Updated title per CodeRabbit's pre-merge check — the original `docs:` prefix was misleading since the changes are implementation work (VIF centering, SES median fallback, GroupShuffleSplit, kernel hyperparameter CV). The new title better reflects the actual code changes.

**Review by TanKhoiTV — APPROVED** [review]
